### ISSUE TITLE
Queries with enabled encryption are incredibly slow

### ISSUE THREAD

**Issue by tjblackheart** [jepiqueau/capacitor-data-storage-sqlite#35]
**Describe the bug**
Hi @jepiqueau , I am looking for an encrypted key/value store for capacitor and found your plugin. While implementing and using this without any encryption works fine, it gets really slow when I enable the encryption and use it with more than one table on the same DB (not tested with just one table though). 

All of this relates to an Android build and running it on a real device in debug mode. "Slow" here means a timeframe between 15 and 30 seconds before the write process finishes.

**To Reproduce**
Have a look at my store implementation below.

**Expected behavior**
The query speed is not (or barely) affected when enabling encryption.

**Screenshots**
No screenshots, but this is my current implementation:

```
// left out type imports.

import { Plugins } from '@capacitor/core'
import 'capacitor-data-storage-sqlite'
const Storage = Plugins.CapacitorDataStorageSqlite

class EntityStorage {
  private _initialized: boolean
  private _table: string

  constructor(table: string) {
    if (!this._initialized) {
      this.init(table)
    }
  }

  private async init(table: string) {
    try {
      this._table = table
      await Storage.openStore({
        database: 'my_app_name',
        table: table,
        encrypted: true,
        mode: 'secret' 
      })
      this._initialized = true
    } catch (err) {
      console.log(err)
    }
  }

  public async getAll(): Promise<T[]> {
    const data = [] as T[]

    await Storage.setTable({ table: this._table })
    const { values } = await Storage.values()

    for (const value of values) {
      data.push(JSON.parse(value))
    }

    return data
  }

  public async getByKey(key: string): Promise<T | null> {
    await Storage.setTable({ table: this._table })
    const { value } = await Storage.get({ key })
    return (value) ? JSON.parse(value) as T : null
  }

  public async saveItem(value: T) {
    await Storage.setTable({ table: this._table })
    await Storage.set({ key: value.id, value: JSON.stringify(value) })
  }

  public async deleteItem(key: string): Promise<void> {
    await Storage.setTable({ table: this._table })
    await Storage.remove({ key })
  }

  public async clearEntries(): Promise<void> {
    const items = await this.getAll()

    for (const item of items) {
      if (item && item.id) {
        await this.deleteItem(item.id)
      }
    }
  }

  public async clearStorage() {
    await Storage.clear()
  }
}

const aStorage = new class AStorage extends EntityStorage<A> {
  constructor(){ super('a') }
}()

const bStorage = new class BStorage extends EntityStorage<B> {
  constructor() { super('b') }
}()

const cStorage = new class CStorage extends EntityStorage<C> {
  constructor() { super('c') }
}()

export {
  aStorage,
  bStorage,
  cStorage
}

```

**Smartphone (please complete the following information):**

- Device: OnePlus 6T
- OS: Android 10
- Browser: -
- Version: -

**Comment by jepiqueau** [comment]
@tjblackheart First thanks for using the plugin. Yes, i notice too that it is quite slow with Android which is less visible for iOS. I should rewrite the android plugin by using `androidx.sqlite.db` like i did for the `capacitor-community/sqlite` plugin and remove the open and close the database after each set or get operation. This will take me a bit of time to do but it will come.  (i have to prioritize my time)
Looking at your code,  you do not need to  set the Table ` await Storage.setTable({ table: this._table })` in each of your command as it has been set in the `init` method with 
``` 
      this._table = table
      await Storage.openStore({
        database: 'my_app_name',
        table: table,
        encrypted: true,
        mode: 'secret' 
      })
      this._initialized = true
```
this should help a bit
Which framework are you using for your App ? Ionic/Angular, Ionic/React, Ionic/Vue or react or vue ?

**Comment by tjblackheart** [comment]
Hi, thanks for the quick answer! We're using ionic with vue (v3). However, after a deeper code inspection I think this is not only a plugin issue - there are way to many writes happening, so the (expected) overhead of the encrypted storage writes add up and make this thing so slow. I think I have to refactor quite a bit and do the DB writes not on each data change, also move them to the background, perhaps only on an app lifecycle change. Thanks for taking a look, though, and thanks for providing this plugin. Cheers!

**Comment by jepiqueau** [comment]
@tjblackheart I have started to rewrite the code for Capacitor 3 and accommodate for multiple store instance and so i will eliminate the open and close of the store instance at each call of a method. This will help a lot for the speed, even if the encryption will be still slowing done the process. We have seen huge speed improvement in doing this for the `@capacitor-community/sqlite@refactor` so i expect the same for this plugin. So this will come, Capacitor 3 is still in alpha release right now and i will follow them, so be patient and optimize also your code during that time, it is not a waste of time. Cheers!

**Comment by tjblackheart** [comment]
Sounds good! For the time being, I now write all of the data to just one table as a big JSON blob, and also do that asynchronous in the background via a watcher, so the app store is not totally depending on the DB store anymore. This works fine regarding the speed, so no worries here.

Could you tell me about how/where the encryptoin key is stored on the device? Do you use the keychain API for this? Also, would it be easy to update the plugin to the latest version and keep all data when capacitor goes v3?

**Comment by jepiqueau** [comment]
@tjblackheart As mentioned in the API.md documentation the keys (passphrases) are stored in

```
To define your own "secret" and "newsecret" keys:

in IOS, go to the Pod/Development Pods/jeepqCapacitor/DataStorageSQLite/Global.swift file
in Android, go to jeepq-capacitor/java/com.jeep.plugins.capacitor/cdssUtils/Global.java
and then update the default values before building your app.
```
Moving to capacitor3 will require some changes in the app see `https://capacitorjs.com/docs/v3/updating/3-0`
and certainly the interface to the plugin will sligthly change but i will try to keep the data and the store as they are.

**Comment by jepiqueau** [comment]
@tjblackheart this has been addressed in release 3.0.0-alpha.6 (iOS & Android)
